Deserialization API rework

piodul opened the pull request:
This PR reworks the deserialization API. The new API solves some performance problems that the old one had and makes it possible to improve type safety. As this change will probably affect lots of existing code, the old API is still available but marked as "Legacy" - it's opt-in but only some small changes will be required to use it.

__What are the problems with the current API?__

Currently, when the driver receives response from the database, it immediately deserializes it to `Vec<Row>`, where `Row = Vec<CqlValue>` and `CqlValue` is an enum that is able to represent any CQL type. Then, the user can choose to iterate over this representation, converting rows with the `FromRow` trait. When a row is converted, it usually uses the `FromCqlVal` trait to convert individual column values. This design has the following problems:

- The representation with rows and vecs is very inefficient. It incurs `#rows + 1` allocations just for `Vec`s only. Moreover, some column types will allocate too - e.g. `text` (represented by a `String`), `blob` (represented with `Vec<u8>`), as well as collections and UDTs. Many of those allocations could be perfectly avoided by deserializing data on demand: non-allocating alternatives such as `&[u8]` can be used to represent some types, collections can be consumed via iterators, UDTs can be deserialized directly to structs with the help of derive macros, and the need to allocate and deserialize everything up front can be avoided by iterating over the rows and deserializing them lazily.
- `FromCqlVal` and `FromRow` receive incomplete information about the rows and columns. For example, `CqlValue::List` variant keeps its values in a Vec, so it can't validate the type if the vec is empty. A more issue with `FromRow` is that it doesn't know about the names of the columns, so the `FromRow` derive macro can't validate that the column types match, nor it can support deserializing structs where fields are in a different order than columns.

Those issues couldn't really be fixed without the API redesign, hence that's why this PR have been created.

__The new API__

The new API is based on two new traits, `DeserializeCql` and `DeserializeRow`:

```rust
pub trait DeserializeCql<'frame>
where
    Self: Sized,
{
    /// Checks whether this type can be deserialized from given CQL type.
    fn type_check(typ: &ColumnType) -> Result<(), ParseError>;

    /// Deserialize from given bytes.
    /// The function may assume that `type_check` was called and it succeeded.
    fn deserialize(
        typ: &'frame ColumnType,
        v: Option<FrameSlice<'frame>>,
    ) -> Result<Self, ParseError>;
}
```

```rust
pub trait DeserializeRow<'frame>
where
    Self: Sized,
{
    /// Checks whether this type can be deserialized from given row.
    fn type_check(specs: &[ColumnSpec]) -> Result<(), ParseError>;

    /// Deserialize from given bytes.
    /// The function may assume that `type_check` was called and it succeeded.
    fn deserialize(row: ColumnIterator<'frame>) -> Result<Self, ParseError>;
}
```

Instead of receiving pre-parsed data, both traits receive objects that borrow data from the serialized frame - hence the `'frame` lifetime parameter. This allows the types being deserialized to allocate only when they need to (e.g. in case of `String`). However, the new API also allows types that borrow data from the serialized frame - for example `&[u8]` or `&str`. There is also a third option - the type being deserialized can also share ownership of the serialized frame - for example, you can deserialize `blob` to a `bytes::Bytes`, which will keep the original, serialized frame alive while it is alive. This is easier to use than slices or `str`s, but one needs to be careful as even a tiny blob will keep the memory for the whole serialized frame alive, which can lead to space leaks.

Another notable changes is the `type_check` method. In case of `DeserializeRow` it receives more comprehensive information than `FromRow`, which include column names, and now maching column names to fields is possible. The derive for `DeserializeRow` has that functionality. Moreover, because type checking is separate from `deserialize`, it can be done only once when a response is received - if the page with response has 1000 rows then we only need to call `type_check` once per page but `deserialize` once per row. `deserialize` can assume that `type_check` has been called and it can avoid some potentially expensive type checks.

__Migration from the legacy API__

Apart from the use of different traits, the existing API is not that different. Some method names in `QueryResult` and `RowIterator` (now `RawIterator`) have been changed and some fields have been hidden, but otherwise changes that need to be made are frequently mechanical - the commits that adjust the examples and tests are an example of that. Nonetheless, the old API have been existing for a long time, so requiring folks to switch everything at once wouldn't be right in my opinion. This PR also made sure to preserve the old API - while it's opt-in it's very easy to adjust the existing code to use it - just do `use scylla::LegacySession as Session` and similarly for other types that have been marked as "legacy". There are also some facilities that make it possible to migrate the code piecewise - for example, you can create a new API Session from a LegacySession which, although has a different API, shares the same resources. There will also be a migration guide include (although it's still a TODO at this point, one of the reasons that it is a draft PR).

__TODOs__

- [x] Finish docstrings for the most important traits/types that were introduced
- [x] Update documentation and write a migration guide

----

Fixes: https://github.com/scylladb/scylla-rust-driver/issues/462

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented:
There is a typo in `scylla-cql/types: introduce new deserialization traits` commit description: `DeserializeCql<'s>` instead of `DeserializeCql<'f>`.

piodul commented:
v0.1: rebased and fixed clippy's complaints

piodul commented:
v1:
    - Renamed `LegacyXYZ` stuff to `Legacy08XYZ` - if we have more legacy things in the future then it will be good to differentiate them somehow,
    - Adjusted existing documentation to the new API - the code examples now compile,
    - Added migration guide from 0.8 deserialization API,
    - Added documentation for the new derive macros,
    - Added the `no_field_name_verification` annotation for the derive macros,
    - Added docstrings to some of the missing items, slightly moved around some code,
    - Rebased on newest `main`.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+// Implements tuple deserialization.
+// The generated impl expects that the serialized data will contain at least
+// the given amount of values.
+// TODO: Add more context to errors
```
What do you mean here? Please specify in the comment.

piodul replied:
Which part of the comment is unclear?

piodul replied:
Clarified: `TODO: Include information about the id of the column that failed to parse`

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+/// # Note
+///
+/// A serialized UDT will generally have one value for each field, but it is
+/// allowed to have less. This iterator differentiates null values
```
less -> fewer  (_value_ is countable)

piodul replied:
Done

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+    }
+}
+
+impl<'frame> Iterator for UdtIterator<'frame> {
+    type Item = Result<
+        (
+            &'frame (String, ColumnType),
+            Option<Option<FrameSlice<'frame>>>,
+        ),
+        ParseError,
+    >;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // TODO: Should we fail when there are too many fields?
```
Let's discuss this. Should we?

If we got more fields than we expected for a UDT, why should we let it silently pass? IMO it deserves at least a warning. Failure is appropriate as well, as I can't find any counterexample.

piodul replied:
The protocolspec only allows less fields to be sent, and the description implies that it's incorrect to send more:

```
  A UDT value is composed of successive [bytes] values, one for each field of the UDT
  value (in the order defined by the type). A UDT value will generally have one value
  for each field of the type it represents, but it is allowed to have less values than
  the type has fields.
```

The TODO is mostly about considering how to handle a case where the database sends a response that does not conform to the spec. If we don't trust the DB then this method will have to be written a bit differently.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+        compat_check::<Option<i32>>(&ColumnType::Int, serialize_cell(&123i32));
+        compat_check::<Option<i32>>(&ColumnType::Int, make_null());
+
+        // empty values
+        // TODO: FromCqlVal doesn't have support for this
```
Please clarify: what must be done in order to test empty values here?

piodul replied:
Eh, they probably just need a separate test. I don't think we have an equivalent to the new `MaybeEmpty`, just the `CqlValue::Empty`, so `compat_check` can't really be used here.

We just need a test for `MaybeEmpty` - there isn't any at the moment. I'll add a test for it.

piodul replied:
Done

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+        b.put_i32(-1);
+    }
+
+    // TODO: Tests for iterators
```
Just a reminder to make an issue out of this in the future.

wprzytula reviewed (commented):
So far perused only 4 commits, but thanks to their volume it's enough for a firm burst of comments. I'll continue review in the next days.

piodul commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,307 @@
+    ///
+    /// This method is exposed to allow writing deserialization tests
+    /// for custom types.
+    #[inline]
```
If a function is marked `#[inline]`, the compiler is permitted to inline it if it is used from a different crate. If it's not marked like that then it won't be inlined unless LTO is enabled or the function is generic.

I added the `#[inline]` annotation for small functions that will most likely be called by user code. The annotation is still a hint and not a requirement for the compiler, but e.g. in case of trivial getters inlining them should be a win pretty much always.

wprzytula replied:
I wasn't aware that `#[inline]` enables inlining through crate boundary without LTO enabled. GTK.

piodul commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,307 @@
+        self.mem.is_empty()
+    }
+
+    /// Returns a reference to the Bytes object which encompasses the slice.
+    ///
+    /// The Bytes object will usually be larger than the slice returned by
+    /// [FrameSlice::as_slice]. If you wish to obtain a new Bytes object that
+    /// points only to the subslice represented by the FrameSlice object,
+    /// see [FrameSlice::to_bytes].
+    #[inline]
+    pub fn as_bytes_ref(&self) -> &'frame Bytes {
+        self.frame
+    }
```
Yes.

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+    };
+}
+
+impl_strict_type!(
+    "boolean",
+    bool,
+    ColumnType::Boolean,
+    |_typ: &'frame ColumnType, v: Option<FrameSlice<'frame>>| {
+        let val = ensure_not_null(v)?;
+        let arr = ensure_exact_length::<1>("boolean", val)?;
+        Ok(arr[0] != 0x00)
+    }
+);
```
They are reworked - implementation is slightly different, but they should work the same (at least this was the intention). Fortunately, most of those implementations are trivial.

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+            $('frame: $l)?
+        {
+            fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+                // TODO: Print the CQL type name from ColumnType
+                // in the notation that Cassandra uses
```
Yes, the TODO is actually about including the CQL type in the error string.

piodul replied:
Clarified the comment.

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+    }
+);
+
+// TODO: Deserialization for string::String<Bytes>
```
It was originally proposed in https://github.com/scylladb/scylla-rust-driver/issues/571 . I'm not 1000% sure we want to support it, but technically we can. We can decide on that later.

wprzytula commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,307 @@
+//! # use scylla_cql::frame::response::result::ColumnType;
+//! # use scylla_cql::frame::frame_errors::ParseError;
+//! struct MySlice<'a>(&'a [u8]);
+//! impl<'a, 'frame> DeserializeCql<'frame> for MySlice<'a>
+//! where
+//!     'frame: 'a,
+//! {
```
At least for me, the more consise form in this case is better parseable. But it's probable that for others the `where` clause stands out more.

piodul commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.remaining, Some(self.remaining))
+    }
```
Well, it's just inconsistency from my side. It should return `0` as a lower bound (because an error can happen during deserialization).

piodul replied:
Done.

wprzytula commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,307 @@
+    /// This method is exposed to allow writing deserialization tests
+    /// for custom types.
+    #[inline]
+    pub fn new(frame: &'frame Bytes) -> Self {
+        Self { mem: frame, frame }
+    }
+
+    /// Creates a new FrameSlice that refers to a subslice of a given Bytes object.
+    #[inline]
+    pub fn new_subslice(mem: &'frame [u8], frame: &'frame Bytes) -> Self {
+        Self { mem, frame }
+    }
```
Could we then change the totally generic name `mem` into, e.g., `subframe`, `frame_part` or alike?

piodul replied:
I changed `mem` to `frame_subslice` and `frame` to `original_frame`.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+    ) -> Result<Self, ParseError>;
+}
+
+impl<'frame> DeserializeCql<'frame> for CqlValue {
+    fn type_check(_typ: &ColumnType) -> Result<(), ParseError> {
+        // Type checking is done in deserialize
+        Ok(())
+    }
```
Yes, please change it as suggested.

piodul commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,387 @@
+
+/// Represents a raw, unparsed column value.
+#[non_exhaustive]
+pub struct RawColumn<'frame> {
+    pub spec: &'frame ColumnSpec,
+    pub slice: Option<FrameSlice<'frame>>,
+}
```
I'll add the getters, it shouldn't be a problem to do this. Frankly speaking, I don't see how the struct layout could change and how we could benefit from hiding the fields behind getters, but I'd also err on the side of caution.

piodul commented:
@cvybhu @wprzytula review ping

piodul commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -373,6 +377,145 @@ impl Row {
 }
 
+/// Rows response, in partially serialized form.
+// TODO: We could provide ResultMetadata in a similar, lazily
+// deserialized form - now it can be a source of allocations
```
I'd rather not do it in this PR. It's quite complicated already. I'll create an issue after this is merged.

piodul commented:
v1.1: rebased on newest main (no other changes)

cvybhu commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+
+// tuples
+
+// Implements tuple deserialization.
+// The generated impl expects that the serialized data will contain at least
+// the given amount of values.
```
I think that for tuples we should be checking that the number of values is exactly the same. With UDTs the situation is different because it's possible to add new fields to them, but AFAIK it's impossible to add a new field to a tuple.

piodul replied:
Ok, that used to be the case in the old framework, so I'll bring this behavior back.

I wonder whether it's possible to add/remove fields in a tuple column. If it is, then we should consider adding types that aren't as strict and e.g. allow unknown fields at the end.

piodul replied:
Actually, I realized that this comment is wrong. The generated `type_check` checks that the tuple has the right size. I'll adjust this comment.

cvybhu reviewed (changes requested):
Looks very nice, I got through some of it, will continue later.

cvybhu commented on scylla-macros/src/deserialize/row.rs:
```diff
@@ -0,0 +1,500 @@
+            {
+                let col = row.next().ok_or_else(|| {
+                    #crate_path::frame::frame_errors::ParseError::BadIncomingData(
+                        "Not enough fields".to_owned(),
```
This message is misleading. When `ok_or_else` fails that means that there are more fields than rows in the `ColumnIterator`.

piodul replied:
I'll change it to "Not enough columns", then.

cvybhu commented on scylla-macros/src/deserialize/row.rs:
```diff
@@ -0,0 +1,500 @@
+                        unknown => {
+                            return ::std::result::Result::Err(
+                                #crate_path::frame::frame_errors::ParseError::BadIncomingData(
+                                    ::std::format!("Unknown field: {}", unknown),
```
Maybe "unexpected column: {}" would be better

cvybhu commented on scylla-macros/src/deserialize/row.rs:
```diff
@@ -0,0 +1,500 @@
+                    #(#append_name_blocks)*
+                    return ::std::result::Result::Err(
+                        #crate_path::frame::frame_errors::ParseError::BadIncomingData(
+                            ::std::format!("Missing fields: {:?}", missing_fields),
```
`Missing columns`?

cvybhu commented on scylla-macros/src/deserialize/row.rs:
```diff
@@ -0,0 +1,500 @@
+            let cql_name_literal = field.cql_name_literal();
+            parse_quote!(#deserialize_field.ok_or_else(|| {
+                #crate_path::frame::frame_errors::ParseError::BadIncomingData(
+                    ::std::format!("Missing field: {}", #cql_name_literal)
```
`Missing column`?

cvybhu commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1801 @@
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError>;
+}
+
```
I'm a bit worried that the user has to remember to call `type_check` before doing `deserialize`. As the code gets more complex, with all the iterators and whatnot it's hard to prove to myself that `type_check` was indeed called.

I wonder if it would be possible to use the type system to make such mistakes impossible. Something like `TypeChecked<T>()` or something.

piodul replied:
I don't think it is a problem for the users at all. We, as the developers, need to provide abstractions that always call `type_check` before `deserialize` and the users will be fine - and this PR provides them. Apart from implementing `DeserializeCql/DeserializeRow` themselves, I don't think there will ever be a need for the users to call `type_check` or `deserialize` directly.

cvybhu commented on scylla/src/transport/iterator.rs:
```diff
@@ -74,11 +74,19 @@ macro_rules! ready_some_ok {
+/// Before the results can be processed, the RawIterator needs to be cast
+/// into a typed iterator.
+///
+/// TODO: How?
```
TODO in public documentation

Gor027 commented on scylla/src/transport/query_result.rs:
```diff
@@ -0,0 +1,210 @@
+    /// Fails when the query isn't of a type that could return rows, same as [`rows()`](QueryResult::rows).
+    pub fn maybe_first_row<'s, R: DeserializeRow<'s>>(&'s self) -> Result<Option<R>, RowsError> {
+        Ok(self.rows::<R>()?.next().transpose()?)
+    }
```
Shouldn't `FirstRowError` be returned from `maybe_first_row` function?

piodul replied:
`FirstRowError` - no, it is too broad, it can represent some failures which cannot happen with `maybe_first_row`.

Maybe you meant to say that `maybe_first_row` should return an error if the response was not of Rows type, like the comment suggests? In that case I agree and will fix that.

piodul replied:
I will add a `MaybeFirstRowError` struct.

piodul commented:
v1.2: rebased, fixed one test that used to fail in serverless mode

piodul commented on scylla/src/transport/legacy_query_result.rs:
```diff
@@ -10,7 +10,7 @@ use uuid::Uuid;
 #[non_exhaustive]
 #[derive(Default, Debug)]
-pub struct QueryResult {
+pub struct Legacy08QueryResult {
```
This should actually be fixed now, I unintentionally included that in the rebase.

piodul commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,387 @@
+                    )*
+                    return Ok(());
+                }
+                const TUPLE_LEN: usize = [0, $($idx),*].len() - 1;
```
Credit goes to @havaker who found it first: https://github.com/scylladb/scylla-rust-driver/pull/288/files#diff-c9cfc556f95e6a898e094c2f062cbde41f22a8e6781687defc010814e159f24aR131

piodul commented on scylla-cql/Cargo.toml:
```diff
@@ -28,6 +28,7 @@ serde = { version = "1.0", optional = true }
 
```
Fixed

piodul commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,307 @@
+//! which point to a subslice but keep the whole, original `Bytes` object alive.
+//!
+//! During deserialization, a type can obtain a `Bytes` subslice that point
+//! to the serialized value. This approach combined advantages of the previous
```
Fixed

piodul commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,307 @@
+//! of the serialized response. It supports creating derived `Bytes` objects
+//! which point to a subslice but keep the whole, original `Bytes` object alive.
+//!
+//! During deserialization, a type can obtain a `Bytes` subslice that point
```
Fixed

piodul commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,307 @@
+pub struct FrameSlice<'frame> {
+    mem: &'frame [u8],
+    frame: &'frame Bytes,
+}
```
Done, I added some explanatory comments.

piodul commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,387 @@
+        <R as DeserializeRow<'frame>>::deserialize(iter)
+    }
+
+    // TODO: Row iterator tests
```
I went ahead and implemented some tests for RowIterator.

piodul commented:
v2.1: rebased on main

piodul commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -373,6 +377,145 @@ impl Row {
+        Some(Ok(iter))
+    }
+
+    #[inline]
+    pub fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(self.remaining))
+    }
```
The iterator can produce less rows if it fails to parse. It shouldn't usually be the case, but there is a possibility. If we always returned `self.remaining` as a lower bound, this could be incorrect.

I'm not entirely sure though whether we should specify `self.remaining` or `0` as a lower bound. Although [the documentation](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint) specifies that this is just a hint for optimization purposes, it also states that returning wrong bounds is a "violation of the trait's protocol". I found some discussions which seem to agree with the interpretation that wrong bounds are a bug; the code that calls it can assume that this implementation is correct provided that it doesn't do anything unsafe under that assumption (but it's e.g. perfectly fine to panic) - for example, see here: https://internals.rust-lang.org/t/size-hint-correctness-reproducibility-and-documentation/8058 .

On the other hand, providing a tighter lower bound indeed could lead to better optimization, but I'd rather not risk it because it's not clear to me whether an implementation of `size_hint` that is sometimes incorrect can lead to panics in the caller.

piodul commented on scylla/src/transport/query_result.rs:
```diff
@@ -0,0 +1,210 @@
+    pub fn maybe_rows<'s, R: DeserializeRow<'s>>(
+        &'s self,
+    ) -> Result<Option<TypedRowIterator<'s, R>>, ParseError> {
+        match &self.raw_rows {
+            Some(rows) => Ok(Some(rows.rows_iter()?)),
+            None => Ok(None),
+        }
```
I tend to agree with @cvybhu here. Although I like the functional style myself, combinators like `and_then`, `map`, `transpose` are good only in moderation. Understanding an expression that uses those combinators heavily requires shuffling a lot of context in your head - you need to manually infer the type of the expression after each dot in order to properly understand what it does. This is especially tricky here because of the nesting of `Option` and `Result` layers. Writing such expressions is a lot of fun, and reading them is a nice mental exercise, but I think it's only detrimental in production code.

Actually, the `QueryResult::maybe_first_row` that I wrote also suffers from the same problem - I'll try to make it more understandable.

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+            let v = V::deserialize(self.v_typ, raw_v)?;
+            Ok((k, v))
+        };
+        do_next().map(Some).transpose()
```
Huh. It is. What a ridiculously over-complicated way to wrap a Result value in Some...

piodul commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,307 @@
+    #[inline]
+    pub fn as_bytes_ref(&self) -> &'frame Bytes {
+        self.frame
+    }
```
Technically speaking, it doesn't return a `Bytes` object, only a reference - and I'd keep it that way for performance reasons - cloning and truncating `Bytes` has some cost, and we only need a reference to the original `Bytes`.

Instead, I'd rather change the name of the method. How about `original_frame()` (the `frame` field has been renamed to `original_frame`, btw)?

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+            Ok(MaybeEmpty::Value(v))
+        }
+    }
+}
```
Hmm, I remember that I considered it when implementing `MaybeEmpty`. I think the reason for why I dropped was that it's easy to forget to add `EmptyableCqlType` when adding a new `impl DeserializeCql`. Besides, while `MaybeEmpty<String>` doesn't make much sense from the CQL protocol point of view, it does work and it should be pretty clear to the user what it does.

However, now that I think of it, it isn't a huge problem and adding missing `impl EmptyableCqlType` is easy and doesn't break the API. Moreover, unless users are well-versed with the CQL protocol spec (I wouldn't expect or require them to be), they won't know which types can be empty, so this might serve an informative purpose. In the future, if we rework the serialization API, then we'd rather use `MaybeEmpty` there, too - and the `EmptyableCqlType` will be even more important for the purpose of serialization.

To sum up - I'm convinced, I'll add a trait.

piodul commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,387 @@
+
+/// A typed version of `RowIterator` which deserializes the rows before
+/// returning them.
+pub struct TypedRowIterator<'frame, R> {
```
IMO there is enough context for most people to infer that `R` means "Row": R is the first letter of Row, and the type parameter is a part of the `TypedRowIterator` struct.

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+
+// other numeric types
+
+impl_strict_type!(
+    "varint",
+    num_bigint::BigInt,
+    ColumnType::Varint,
+    |_typ: &'frame ColumnType, v: Option<FrameSlice<'frame>>| {
+        let val = ensure_not_null(v)?;
+        Ok(num_bigint::BigInt::from_signed_bytes_be(val))
+    }
+);
```
I'm not sure. I'd rather not make this decision in this PR and keep the old semantics as closely as possible. Please open an issue if you think we should do this.

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+                    "Value is out of representable range for NaiveDate".to_string(),
+                )
+            })
+    }
```
Isn't it obvious in the new code? The old code operated on `i64`s implicitly, i.e. it subtracted two `chrono::Duration` values. In the new code, we are operating directly on `i64`s.

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1515 @@
+// lists and sets
+
+/// An iterator over either a CQL set or list.
+pub struct SequenceIterator<'frame, T> {
```
I'm not sure I like the term "listlike". It leans too heavily towards "list" and not enough towards "set".

On the other hand, "sequence" is not perfect either as it implies that the elements are ordered (which is not true in case of a set).

piodul commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -373,6 +377,145 @@ impl Row {
+#[derive(Debug)]
+pub struct RawRowsLendingIterator {
+    metadata: ResultMetadata,
+    remaining: usize,
+    at: usize,
```
> IIUC, `at` is related to bytes and `remaining` to row count?

Yes. Will change to `remaining_rows` and `at_byte`.

piodul commented on scylla-cql/src/macros.rs:
```diff
@@ -13,6 +13,167 @@ pub use scylla_macros::IntoUserType;
+///
+/// `#[scylla(default_when_missing)]`
+///
+/// If the UDT definition does not contain this field, it will be initialized
+/// with `Default::default()`. __This attribute has no effect in `enforce_order`
+/// mode.__
```
Ok, we can support `enforce_order + default_when_missing`. If a struct field doesn't match the column name, we will initialize that field to `Default::default()` and will try to initialize the next field with that column.

piodul commented on scylla-macros/src/lib.rs:
```diff
@@ -1,11 +1,30 @@
 mod parser;
 mod value_list;
 
+#[proc_macro_derive(DeserializeRow, attributes(scylla, scylla_crate))]
+pub fn deserialize_row_derive(tokens_input: TokenStream) -> TokenStream {
+    match deserialize::row::deserialize_row_derive(tokens_input) {
+        Ok(tokens) => tokens.into_token_stream().into(),
+        Err(err) => err.into_compile_error().into(),
+    }
+}
+
+#[proc_macro_derive(DeserializeCql, attributes(scylla))]
+pub fn deserialize_cql_derive(tokens_input: TokenStream) -> TokenStream {
+    match deserialize::cql::deserialize_user_type_derive(tokens_input) {
+        Ok(tokens) => tokens.into_token_stream().into(),
+        Err(err) => err.into_compile_error().into(),
+    }
+}
```
It shouldn't, good catch. The `scylla_crate` attribute was used by the old macros, the new ones only use `scylla`.

piodul commented on scylla-macros/Cargo.toml:
```diff
@@ -12,6 +12,7 @@ license = "MIT OR Apache-2.0"
 quote = "1.0" 
-proc-macro2 = "1.0"
\ No newline at end of file
+proc-macro2 = "1.0"
```
Yes, by convention it is better to have a newline at the end. I'm not sure what you meant to imply here as the PR, though unintentionally, does the right thing and adds a newline at the end.

piodul commented on scylla-cql/src/macros.rs:
````diff
@@ -13,6 +13,167 @@ pub use scylla_macros::IntoUserType;
+///
+/// A table defined like this:
+///
+/// ```notrust
+/// CREATE TABLE ks.my_table (a PRIMARY KEY, b text, c blob);
````
Will change to `text`.

wprzytula commented on docs/source/queries/result.md:
```diff
@@ -2,62 +2,13 @@
 
 `Session::query` and `Session::execute` return a `QueryResult` with rows represented as `Option<Vec<Row>>`.
 
```
This appears to no longer be true, as `QueryResult` has only `raw_rows` after the refactor.

wprzytula commented on docs/source/queries/result.md:
```diff
@@ -2,62 +2,13 @@
-* `first_row_typed::<RowT>` - same as `maybe_first_row`, but fails without the first row
-* `single_row_typed::<RowT>` - same as `first_row`, but fails when there is more than one row
+* `rows::<RowT>()` - returns the rows parsed as the given type
+* `maybe_first_row::<RowT>` - returns `Option<RowT>` containing first row from the result
+* `first_row::<RowT>` - same as `maybe_first_row`, but fails without the first row
```
For me, it is not clear how `first_row` fails. Does it panic?

wprzytula commented on docs/source/migration-guides/post-0.8-deserialization.md:
```diff
@@ -0,0 +1,264 @@
+While it's possible to implement those manually, the driver provides procedural macros for automatic derivation in some cases:
+
+- `FromRow` - implements `FromRow` for a struct.
+- `FromUserType` - generated an implementation of `FromCqlVal` for the struct, trying to parse the CQL value as a UDT.
```
s/generated/generates

wprzytula commented on docs/source/migration-guides/post-0.8-deserialization.md:
```diff
@@ -0,0 +1,264 @@
+- `FromRow` - implements `FromRow` for a struct.
+- `FromUserType` - generated an implementation of `FromCqlVal` for the struct, trying to parse the CQL value as a UDT.
+
+Note: the macros above have a default behavior that is different than what `FromRow` and `FromUserTypes` do.
+
```
This is unclear.

wprzytula replied:
What is their default behaviour and what are the differences? Also, as the names here are similar both for macros and traits, it is crucial to distinguish them with some strong markers.

wprzytula commented on docs/source/migration-guides/post-0.8-deserialization.md:
```diff
@@ -0,0 +1,264 @@
+
+### New traits
+
+The new API introduce two analogous traits that, instead of consuming pre-parsed `Vec<Option<CqlValue>>`, are given raw, serialized data with full information about its type. This leads to better performance and allows for better type safety.
+
```
What kind of information about the serialized data is given to the new API? Provide examples.

wprzytula commented on docs/source/migration-guides/post-0.8-deserialization.md:
````diff
@@ -0,0 +1,264 @@
+}
+```
+
+The above traits have been implemented for the same set of types as `FromRow` and `FromCqlVal`, respectively. Notably, `DeserializeRow` is implemented for `Row`, and `DeserializeCql` is implemented for `CqlValue`.
````
It can be helpful to remind that `Row` and `CqlValue` are representatives of the legacy API.

wprzytula reviewed (changes requested):
Apart from the commit introducing new procedural macros, I've completed the review.

piodul commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1528 @@
+        <T as DeserializeCql<'frame>>::deserialize(typ, value)
+    }
+
+    fn make_cell(cell: &[u8]) -> Bytes {
+        let mut b = BytesMut::new();
+        append_cell(&mut b, cell);
+        b.freeze()
+    }
```
I'll change it to "make_bytes" here because it's more in line with the protocol spec. The "cell" term is vaguely similar to a concept of "atomic cell" in ScyllaDB's codebase, but it's only superficially similar, so I'll get rid of it here to reduce confusion.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -373,6 +377,145 @@ impl Row {
+    /// The column iterator must be consumed before the rows iterator can
+    /// continue.
+    #[inline]
+    pub fn next(&mut self) -> Option<StdResult<ColumnIterator, ParseError>> {
```
In more detail, lifetime elision is applicable here because there is only one possible way of inferring the lifetimes: if we return a value with some lifetime ("one lifetime output"), and we provide only one parameter that has a lifetime exceeding this function's scope ("only one lifetime input"), then their lifetimes must match - that's all.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -977,12 +977,10 @@ pub fn deser_cql_value(typ: &ColumnType, buf: &mut &[u8]) -> StdResult<CqlValue,
 
-    let original_size = buf.len();
-
     // TODO: the protocol allows an optimization (which must be explicitly requested on query by
     // the driver) where the column metadata is not sent with the result.
     // Implement this optimization. We'll then need to take the column types by a parameter.
```
Are you going to change it in this PR?

Gor027 commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,1528 @@
+}
+
+/// An iterator over a CQL map.
+pub struct MapIterator<'frame, K, V> {
+    k_typ: &'frame ColumnType,
+    v_typ: &'frame ColumnType,
+    raw_iter: FixedLengthBytesSequenceIterator<'frame>,
+    phantom_data_k: std::marker::PhantomData<K>,
+    phantom_data_v: std::marker::PhantomData<V>,
+}
```
I think `MapIterator` should have a public constructor like `SequenceIterator` or `UdtIterator`. This will be also helpful for the new API integration in cpp-rust bindings.

wprzytula reviewed (changes requested):
The new derive macros' code looks promising. I've also examined the generated code using `cargo expand` and it looks valid, too. Only minor nits there.

piodul commented:
I'll split this PR into multiple smaller ones. It's quite hard to keep going back to a single huge PR, rebasing it and addressing lots of comments at once, so it should make the process smoother. I'll address the existing comments in the upcoming, smaller PRs.
